flax/experimental/nnx/tests/test_graph_utils.py

```diff
@@ -277,3 +278,116 @@ def __init__(self):
     assert m2.tree.b == 'a'
     assert m2.tree.a is not m.tree.a
     assert m2.tree is not m.tree
+
+  def test_cached_unflatten(self):
+    class Foo(nnx.Module):
+      def __init__(self, *, rngs: nnx.Rngs):
+        self.a = nnx.Linear(2, 2, rngs=rngs)
+        self.b = nnx.BatchNorm(2, rngs=rngs)
+
+    def f(m: Foo):
+      m.a, m.b = m.b, m.a
+
+    m = Foo(rngs=nnx.Rngs(0))
+    a = m.a
+    b = m.b
+
+    static: nnx.graph_utils.GraphDef[Foo]
+    state, static, ref_out_idx_out = nnx.graph_utils.graph_flatten(m)
+
+    @partial(jax.jit, static_argnums=(0,))
+    def f_pure(static: nnx.graph_utils.GraphDef[Foo], state):
+      m, idx_out_ref_in = nnx.graph_utils.graph_unflatten(static, state, None)
+      f(m)
+      state, static, ref_in_idx_in = nnx.graph_utils.graph_flatten(m)
+      idx_out_idx_in = nnx.graph_utils.compose_mapping(
+        idx_out_ref_in, ref_in_idx_in
+      )
+      static_out = nnx.graph_utils.Static((static, idx_out_idx_in))
+      return state, static_out
+
+    static_out: nnx.graph_utils.Static
+    state, static_out = f_pure(static, state)
+    idx_out_idx_in: dict[int, int]
+    static, idx_out_idx_in = static_out.value
+    idx_in_ref_out = nnx.graph_utils.compose_mapping(
+      ref_out_idx_out, idx_out_idx_in, invert=True
+    )
+    m2, _ = nnx.graph_utils.graph_unflatten(static, state, idx_in_ref_out)
+    assert m2 is m
+    assert m2.a is b
+    assert m2.b is a
+
+  def test_cached_unflatten_swap_variables(self):
+    class Foo(nnx.Module):
+      def __init__(self):
+        self.a = nnx.Param(1)
+        self.b = nnx.Param(2)
+
+    def f(m: Foo):
+      m.a, m.b = m.b, m.a
+
+    m = Foo()
+    a = m.a
+    b = m.b
+
+    static: nnx.graph_utils.GraphDef[Foo]
+    state, static, ref_out_idx_out = nnx.graph_utils.graph_flatten(m)
+
+    @partial(jax.jit, static_argnums=(0,))
+    def f_pure(static: nnx.graph_utils.GraphDef[Foo], state):
+      m, idx_out_ref_in = nnx.graph_utils.graph_unflatten(static, state, None)
+      f(m)
+      state, static, ref_in_idx_in = nnx.graph_utils.graph_flatten(m)
+      idx_out_idx_in = nnx.graph_utils.compose_mapping(
+        idx_out_ref_in, ref_in_idx_in
+      )
+      static_out = nnx.graph_utils.Static((static, idx_out_idx_in))
+      return state, static_out
+
+    static_out: nnx.graph_utils.Static
+    state, static_out = f_pure(static, state)
+    idx_out_idx_in: dict[int, int]
+    static, idx_out_idx_in = static_out.value
+    idx_in_ref_out = nnx.graph_utils.compose_mapping(
+      ref_out_idx_out, idx_out_idx_in, invert=True
+    )
+    m2, _ = nnx.graph_utils.graph_unflatten(static, state, idx_in_ref_out)
+    assert m2 is m
+    assert m2.a is b
+    assert m2.b is a
+
+  def test_cached_unflatten_add_self_reference(self):
+    class Foo(nnx.Module):
+      def __init__(self):
+        self.ref = None
+
+    def f(m: Foo):
+      m.ref = m
+
+    m = Foo()
+
+    static: nnx.graph_utils.GraphDef[Foo]
```

Comment:
Nit: I'd expect the type checker to be able to infer this type with no hints.

If you decide to remove this declaration, please remove other similar ones in this file as well.